File: Breadth-First Search/Topological sort/LET_no1306_Jump Game III.py
class Solution1:
    def canReach(self, arr: List[int], start: int) -> bool:
        if 0 <= start < len(arr) and arr[start] >= 0:
            arr[start] = -arr[start]
            return arr[start] == 0 or self.canReach(arr, start + arr[start]) or self.canReach(arr, start - arr[start])
        return False

class Solution2:
    def canReach(self, arr: List[int], start: int) -> bool:
        queue = collections.deque([start])
        n = len(arr)
        while queue:
            cur = queue.popleft()
            if arr[cur] == 0:
                return True
            for i in [1, -1]:
                nxt = cur + i * arr[cur]
                if 0 <= nxt < n and arr[nxt] >= 0:
                    arr[nxt] = -arr[nxt]
                    queue.append(nxt)
        return False

File: Breadth-First Search/Topological sort/LET_no1306_Jump Game III.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        queue<int> q;
        q.push(start);
        vector<int> c = {-1, 1};
        while (!q.empty()) {
            int cur = q.front();
            q.pop();
            if (arr[cur] == 0)
                return true;
            for (int i : c) {
                int nxt = cur + i * arr[cur];
                if (nxt < 0 || nxt >= arr.size() || arr[nxt] < 0)
                    continue;
                arr[nxt] = -arr[nxt];
                q.push(nxt);
            }
        }
        return false;
    }
};